Protobuf compiler 3.10 generates Costing_Name clashing with valhalla::Costing_Name

TL;TR: When compiling Valhalla's `.proto` files using Protobuf **3.10**, the `protoc` generates `Foo_Name` and `Foo_Parse` functions which clash with existing hand-rolled ones of the very same names (and possibly similar purpose).

Let's consider the problem using the [enum Costing in options.proto](https://github.com/valhalla/valhalla/blob/3176db5363b4a3c678532b8d496274858977776d/proto/options.proto#L33) and [Costing_Name](https://github.com/valhalla/valhalla/blob/3176db5363b4a3c678532b8d496274858977776d/src/worker.cc#L952) function for concrete example.

First, the [valhalla namespace in src/worker.cc](https://github.com/valhalla/valhalla/blob/3176db5363b4a3c678532b8d496274858977776d/src/worker.cc#L889) defines the collection of `Foo_Name` and `Foo_Parse` functions, including the `Costing_Name` function.

Second, compilation of the `options.proto` using `protoc` **3.10** produces the C++ files which contain the following

- `options.pb.h`

    ```cpp
    const std::string& Costing_Name(Costing value);

    template<typename T>
    inline const std::string& Costing_Name(T enum_t_value) {
    static_assert(::std::is_same<T, Costing>::value ||
        ::std::is_integral<T>::value,
        "Incorrect type passed to function Costing_Name.");
    return Costing_Name(static_cast<Costing>(enum_t_value));
    }
    ```

- `options.pb.cc`

    ```cpp
    const std::string& Costing_Name(
        Costing value) {
    static const bool dummy =
        ::PROTOBUF_NAMESPACE_ID::internal::InitializeEnumStrings(
            Costing_entries,
            Costing_entries_by_number,
            13, Costing_strings);
    (void) dummy;
    int idx = ::PROTOBUF_NAMESPACE_ID::internal::LookUpEnumName(
        Costing_entries,
        Costing_entries_by_number,
        13, value);
    return idx == -1 ? ::PROTOBUF_NAMESPACE_ID::internal::GetEmptyString() :
                        Costing_strings[idx].get();
    }
    ```

among other `Foo_Parse` and `Foo_Name` definitions.

This leads to the redefinition and link-error, for example, when linking tools against static archive `valhalla.lib`:

```console
valhalla.lib(worker.obj) : 
  error LNK2005: "class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > const & __cdecl valhalla::Costing_Name(enum valhalla::Costing)" 
    (?Costing_Name@valhalla@@YAAEBV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@W4Costing@1@@Z) 
  already defined in valhalla.lib(options.pb.obj)
  [C:\projects\valhalla\build\valhalla_run_isochrone.vcxproj]
```

Compilation of `options.proto` with `protoc` **3.2** does not output any such `Foo_Parse` and `Foo_Name` definitions, so no link-error occurs.

The [valhalla-protoc-3.2-vs-3.10.zip](https://github.com/valhalla/valhalla/files/3719101/valhalla-protoc-3.2-vs-3.10.zip) contains full versions of the `protoc`-generated files for comparison:

```
valhalla-protoc-3.10-options.pb.cc
valhalla-protoc-3.10-options.pb.h
valhalla-protoc-3.2-options.pb.cc
valhalla-protoc-3.2-options.pb.h
```

My experience with Protobuf is running short to judge if this is something expected or a compatibility breakage.

------

Currently, Valhalla's CI on AppVeyor (and Azure Pipelines) configurations use the Protobuf **3.10**, but since no tools are currently built, just `valhalla.lib`, the link-error does not occur.
I am preparing PR to add `-DENABLE_TOOLS=ON` (and perhaps `-DENABLE_DATA_TOOLS=ON` too) and that attempt revealed the problem.

--- a/README.md
+++ b/README.md
@@ -91,15 +91,17 @@ To install on a Debian or Ubuntu system you need to install its dependencies wit
 ```bash
 sudo add-apt-repository -y ppa:valhalla-core/valhalla
 sudo apt-get update
-sudo apt-get install -y cmake make libtool pkg-config g++ gcc jq lcov protobuf-compiler vim-common libboost-all-dev libboost-all-dev libcurl4-openssl-dev zlib1g-dev liblz4-dev libprime-server0.6.3-dev libprotobuf-dev prime-server0.6.3-bin nodejs npm
+sudo apt-get install -y cmake make libtool pkg-config g++ gcc jq lcov protobuf-compiler vim-common locales libboost-all-dev libboost-all-dev libcurl4-openssl-dev zlib1g-dev liblz4-dev libprime-server-dev libprotobuf-dev prime-server-bin
 #if you plan to compile with data building support, see below for more info
 sudo apt-get install -y libgeos-dev libgeos++-dev liblua5.2-dev libspatialite-dev libsqlite3-dev lua5.2 wget
-if [[ $(grep -cF xenial /etc/lsb-release) > 0 ]]; then sudo apt-get install -y libsqlite3-mod-spatialite; fi
+source /etc/lsb-release
+if [[ $(python -c "print int($DISTRIB_RELEASE > 15)") > 0 ]]; then sudo apt-get install -y libsqlite3-mod-spatialite; fi
 #if you plan to compile with python bindings, see below for more info
 sudo apt-get install -y python-all-dev
-#if you plan to compile with node bindings, run
-curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.33.11/install.sh | bash
-nvm use install 10 && nvm use 10 # must use node 8.11.1 and up because of N-API
+#if you plan to compile with node bindings
+curl -o- https://raw.githubusercontent.com/creationix/nvm/v0.33.11/install.sh | bash #follow instructions in output to use nvm without starting a new bash session
+nvm install 10
+nvm use 10
 npm install --ignore-scripts
 ```
 

--- a/src/loki/matrix_action.cc
+++ b/src/loki/matrix_action.cc
@@ -89,10 +89,10 @@ void loki_worker_t::init_matrix(Api& request) {
 void loki_worker_t::matrix(Api& request) {
   init_matrix(request);
   auto& options = *request.mutable_options();
-  auto costing_name = Costing_Name(options.costing());
+  auto costing_name = Costing_Enum_Name(options.costing());
 
   if (costing_name == "multimodal") {
-    throw valhalla_exception_t{140, Options_Action_Name(options.action())};
+    throw valhalla_exception_t{140, Options_Action_Enum_Name(options.action())};
   };
 
   // check that location size does not exceed max.

--- a/src/loki/route_action.cc
+++ b/src/loki/route_action.cc
@@ -58,7 +58,7 @@ void loki_worker_t::init_route(Api& request) {
 void loki_worker_t::route(Api& request) {
   init_route(request);
   auto& options = *request.mutable_options();
-  auto costing_name = Costing_Name(options.costing());
+  auto costing_name = Costing_Enum_Name(options.costing());
   check_locations(options.locations_size(), max_locations.find(costing_name)->second);
   check_distance(options.locations(), max_distance.find(costing_name)->second);
 

--- a/src/loki/trace_route_action.cc
+++ b/src/loki/trace_route_action.cc
@@ -155,9 +155,9 @@ void loki_worker_t::init_trace(Api& request) {
 
 void loki_worker_t::trace(Api& request) {
   init_trace(request);
-  auto costing = Costing_Name(request.options().costing());
+  auto costing = Costing_Enum_Name(request.options().costing());
   if (costing == "multimodal") {
-    throw valhalla_exception_t{140, Options_Action_Name(request.options().action())};
+    throw valhalla_exception_t{140, Options_Action_Enum_Name(request.options().action())};
   };
 }
 

--- a/src/loki/worker.cc
+++ b/src/loki/worker.cc
@@ -68,7 +68,7 @@ void loki_worker_t::parse_costing(Api& api) {
   }
 
   auto costing_type = options.costing();
-  auto costing_str = Costing_Name(costing_type);
+  auto costing_str = Costing_Enum_Name(costing_type);
 
   if (!options.do_not_track()) {
     valhalla::midgard::logging::Log("costing_type::" + costing_str, " [ANALYTICS] ");
@@ -155,7 +155,7 @@ loki_worker_t::loki_worker_t(const boost::property_tree::ptree& config,
   Options::Action action;
   for (const auto& kv : config.get_child("loki.actions")) {
     auto path = kv.second.get_value<std::string>();
-    if (!Options_Action_Parse(path, &action)) {
+    if (!Options_Action_Enum_Parse(path, &action)) {
       throw std::runtime_error("Action not supported " + path);
     }
     action_str.append("'/" + path + "' ");

--- a/src/meili/valhalla_run_map_match.cc
+++ b/src/meili/valhalla_run_map_match.cc
@@ -45,7 +45,7 @@ int main(int argc, char* argv[]) {
   rapidjson::read_json(argv[1], config);
   const std::string modename = config.get<std::string>("meili.mode");
   valhalla::Costing costing;
-  if (!valhalla::Costing_Parse(modename, &costing)) {
+  if (!valhalla::Costing_Enum_Parse(modename, &costing)) {
     throw std::runtime_error("No costing method found");
   }
 

--- a/src/sif/transitcost.cc
+++ b/src/sif/transitcost.cc
@@ -692,7 +692,7 @@ void ParseTransitCostOptions(const rapidjson::Document& doc,
     auto filter_stop_action_str =
         rapidjson::get_optional<std::string>(*json_costing_options, "/filters/stops/action");
     FilterAction filter_stop_action;
-    if (filter_stop_action_str && FilterAction_Parse(*filter_stop_action_str, &filter_stop_action)) {
+    if (filter_stop_action_str && FilterAction_Enum_Parse(*filter_stop_action_str, &filter_stop_action)) {
       pbf_costing_options->set_filter_stop_action(filter_stop_action);
       // filter_stop_ids
       auto filter_stop_ids_json =
@@ -710,7 +710,7 @@ void ParseTransitCostOptions(const rapidjson::Document& doc,
         rapidjson::get_optional<std::string>(*json_costing_options, "/filters/operators/action");
     FilterAction filter_operator_action;
     if (filter_operator_action_str &&
-        FilterAction_Parse(*filter_operator_action_str, &filter_operator_action)) {
+        FilterAction_Enum_Parse(*filter_operator_action_str, &filter_operator_action)) {
       pbf_costing_options->set_filter_operator_action(filter_operator_action);
       // filter_operator_ids
       auto filter_operator_ids_json =
@@ -728,7 +728,7 @@ void ParseTransitCostOptions(const rapidjson::Document& doc,
         rapidjson::get_optional<std::string>(*json_costing_options, "/filters/routes/action");
     FilterAction filter_route_action;
     if (filter_route_action_str &&
-        FilterAction_Parse(*filter_route_action_str, &filter_route_action)) {
+        FilterAction_Enum_Parse(*filter_route_action_str, &filter_route_action)) {
       pbf_costing_options->set_filter_route_action(filter_route_action);
       // filter_route_ids
       auto filter_route_ids_json =

--- a/src/sif/transitcost.cc
+++ b/src/sif/transitcost.cc
@@ -692,7 +692,8 @@ void ParseTransitCostOptions(const rapidjson::Document& doc,
     auto filter_stop_action_str =
         rapidjson::get_optional<std::string>(*json_costing_options, "/filters/stops/action");
     FilterAction filter_stop_action;
-    if (filter_stop_action_str && FilterAction_Enum_Parse(*filter_stop_action_str, &filter_stop_action)) {
+    if (filter_stop_action_str &&
+        FilterAction_Enum_Parse(*filter_stop_action_str, &filter_stop_action)) {
       pbf_costing_options->set_filter_stop_action(filter_stop_action);
       // filter_stop_ids
       auto filter_stop_ids_json =

--- a/src/thor/matrix_action.cc
+++ b/src/thor/matrix_action.cc
@@ -29,7 +29,7 @@ std::string thor_worker_t::matrix(Api& request) {
   const auto& options = request.options();
 
   if (!options.do_not_track()) {
-    valhalla::midgard::logging::Log("matrix_type::" + Options_Action_Name(options.action()),
+    valhalla::midgard::logging::Log("matrix_type::" + Options_Action_Enum_Name(options.action()),
                                     " [ANALYTICS] ");
   }
 

--- a/src/thor/trace_attributes_action.cc
+++ b/src/thor/trace_attributes_action.cc
@@ -68,7 +68,7 @@ std::string thor_worker_t::trace_attributes(Api& request) {
         map_match_results.emplace_back(1.0f, 0.0f, std::vector<thor::MatchResult>{});
       } catch (const std::exception& e) {
         throw valhalla_exception_t{
-            443, ShapeMatch_Name(options.shape_match()) +
+            443, ShapeMatch_Enum_Name(options.shape_match()) +
                      " algorithm failed to find exact route match.  Try using "
                      "shape_match:'walk_or_snap' to fallback to map-matching algorithm"};
       }
@@ -80,7 +80,7 @@ std::string thor_worker_t::trace_attributes(Api& request) {
         map_match_results = map_match(request, options.best_paths());
       } catch (const std::exception& e) {
         throw valhalla_exception_t{
-            444, ShapeMatch_Name(options.shape_match()) +
+            444, ShapeMatch_Enum_Name(options.shape_match()) +
                      " algorithm failed to snap the shape points to the correct shape."};
       }
       break;
@@ -93,13 +93,13 @@ std::string thor_worker_t::trace_attributes(Api& request) {
         route_match(request);
         map_match_results.emplace_back(1.0f, 0.0f, std::vector<thor::MatchResult>{});
       } catch (...) {
-        LOG_WARN(ShapeMatch_Name(options.shape_match()) +
+        LOG_WARN(ShapeMatch_Enum_Name(options.shape_match()) +
                  " algorithm failed to find exact route match; Falling back to map_match...");
         try {
           map_match_results = map_match(request);
         } catch (const std::exception& e) {
           throw valhalla_exception_t{
-              444, ShapeMatch_Name(options.shape_match()) +
+              444, ShapeMatch_Enum_Name(options.shape_match()) +
                        " algorithm failed to snap the shape points to the correct shape."};
         }
       }

--- a/src/thor/trace_route_action.cc
+++ b/src/thor/trace_route_action.cc
@@ -76,7 +76,7 @@ void thor_worker_t::trace_route(Api& request) {
         route_match(request);
       } catch (...) {
         throw valhalla_exception_t{
-            443, ShapeMatch_Name(options.shape_match()) +
+            443, ShapeMatch_Enum_Name(options.shape_match()) +
                      " algorithm failed to find exact route match.  Try using "
                      "shape_match:'walk_or_snap' to fallback to map-matching algorithm"};
       }
@@ -96,7 +96,7 @@ void thor_worker_t::trace_route(Api& request) {
       try {
         route_match(request);
       } catch (...) {
-        LOG_WARN(ShapeMatch_Name(options.shape_match()) +
+        LOG_WARN(ShapeMatch_Enum_Name(options.shape_match()) +
                  " algorithm failed to find exact route match; Falling back to map_match...");
         try {
           map_match(request);

--- a/src/thor/worker.cc
+++ b/src/thor/worker.cc
@@ -160,11 +160,12 @@ thor_worker_t::work(const std::list<zmq::message_t>& job,
     double elapsed_time =
         std::chrono::duration<float, std::milli>(std::chrono::system_clock::now() - s).count();
     if (!options.do_not_track() && elapsed_time / denominator > long_request) {
-      LOG_WARN("thor::" + Options_Action_Name(options.action()) +
+      LOG_WARN("thor::" + Options_Action_Enum_Name(options.action()) +
                " request elapsed time (ms)::" + std::to_string(elapsed_time));
-      LOG_WARN("thor::" + Options_Action_Name(options.action()) +
+      LOG_WARN("thor::" + Options_Action_Enum_Name(options.action()) +
                " request exceeded threshold::" + std::to_string(info.id));
-      midgard::logging::Log("valhalla_thor_long_request_" + Options_Action_Name(options.action()),
+      midgard::logging::Log("valhalla_thor_long_request_" +
+                                Options_Action_Enum_Name(options.action()),
                             " [ANALYTICS] ");
     }
 
@@ -212,7 +213,7 @@ std::string thor_worker_t::parse_costing(const Api& request) {
   // Parse out the type of route - this provides the costing method to use
   const auto& options = request.options();
   auto costing = options.costing();
-  auto costing_str = Costing_Name(costing);
+  auto costing_str = Costing_Enum_Name(costing);
 
   // Set travel mode and construct costing
   if (costing == Costing::multimodal || costing == Costing::transit) {
@@ -257,7 +258,7 @@ void thor_worker_t::parse_locations(Api& request) {
       }
 
       // subtract off the min score and cap at max so that path algorithm doesnt go too far
-      auto max_score = kMaxDistances.find(Costing_Name(options.costing()));
+      auto max_score = kMaxDistances.find(Costing_Enum_Name(options.costing()));
       for (auto* candidates : {location.mutable_path_edges(), location.mutable_filtered_edges()}) {
         for (auto& candidate : *candidates) {
           candidate.set_distance(candidate.distance() - minScore);

--- a/src/tyr/matrix_serializer.cc
+++ b/src/tyr/matrix_serializer.cc
@@ -123,7 +123,7 @@ json::MapPtr serialize(const Api& request,
   }
   auto json = json::map({
       {"sources_to_targets", matrix},
-      {"units", Options_Units_Name(options.units())},
+      {"units", Options_Units_Enum_Name(options.units())},
   });
   json->emplace("targets", json::array({locations(options.targets())}));
   json->emplace("sources", json::array({locations(options.sources())}));

--- a/src/tyr/route_serializer_valhalla.cc
+++ b/src/tyr/route_serializer_valhalla.cc
@@ -553,14 +553,14 @@ legs(const google::protobuf::RepeatedPtrField<valhalla::DirectionsLeg>& directio
 
 std::string serialize(const Api& api) {
   // build up the json object
-  auto json =
-      json::map({{"trip", json::map({{"locations", locations(api.directions().routes(0).legs())},
-                                     {"summary", summary(api.directions().routes(0).legs())},
-                                     {"legs", legs(api.directions().routes(0).legs())},
-                                     {"status_message", string("Found route between points")},
-                                     {"status", static_cast<uint64_t>(0)}, // 0 success
-                                     {"units", valhalla::Options_Units_Name(api.options().units())},
-                                     {"language", api.options().language()}})}});
+  auto json = json::map(
+      {{"trip", json::map({{"locations", locations(api.directions().routes(0).legs())},
+                           {"summary", summary(api.directions().routes(0).legs())},
+                           {"legs", legs(api.directions().routes(0).legs())},
+                           {"status_message", string("Found route between points")},
+                           {"status", static_cast<uint64_t>(0)}, // 0 success
+                           {"units", valhalla::Options_Units_Enum_Name(api.options().units())},
+                           {"language", api.options().language()}})}});
   if (api.options().has_id()) {
     json->emplace("id", api.options().id());
   }

--- a/src/tyr/trace_serializer.cc
+++ b/src/tyr/trace_serializer.cc
@@ -500,7 +500,7 @@ std::string serializeTraceAttributes(
 
   // Add units, if specified
   if (request.options().has_units()) {
-    json->emplace("units", valhalla::Options_Units_Name(request.options().units()));
+    json->emplace("units", valhalla::Options_Units_Enum_Name(request.options().units()));
   }
 
   // Loop over all results to process the best path

--- a/src/valhalla_path_comparison.cc
+++ b/src/valhalla_path_comparison.cc
@@ -275,7 +275,7 @@ int main(int argc, char* argv[]) {
   CostFactory<DynamicCost> factory;
   factory.RegisterStandardCostingModels();
   valhalla::Costing costing;
-  if (valhalla::Costing_Parse(routetype, &costing)) {
+  if (valhalla::Costing_Enum_Parse(routetype, &costing)) {
     request.mutable_options()->set_costing(costing);
   } else {
     throw std::runtime_error("No costing method found");

--- a/src/valhalla_run_isochrone.cc
+++ b/src/valhalla_run_isochrone.cc
@@ -160,7 +160,7 @@ int main(int argc, char* argv[]) {
   factory.RegisterStandardCostingModels();
 
   // Get type of route - this provides the costing method to use.
-  std::string routetype = valhalla::Costing_Name(options.costing());
+  std::string routetype = valhalla::Costing_Enum_Name(options.costing());
   LOG_INFO("routetype: " + routetype);
 
   // Get the costing method - pass the JSON configuration

--- a/src/valhalla_run_matrix.cc
+++ b/src/valhalla_run_matrix.cc
@@ -163,7 +163,7 @@ int main(int argc, char* argv[]) {
   factory.RegisterStandardCostingModels();
 
   // Get type of route - this provides the costing method to use.
-  std::string routetype = valhalla::Costing_Name(options.costing());
+  std::string routetype = valhalla::Costing_Enum_Name(options.costing());
   LOG_INFO("routetype: " + routetype);
 
   // Get the costing method - pass the JSON configuration

--- a/src/valhalla_run_route.cc
+++ b/src/valhalla_run_route.cc
@@ -550,7 +550,7 @@ int main(int argc, char* argv[]) {
   const auto& options = request.options();
 
   // Get type of route - this provides the costing method to use.
-  std::string routetype = valhalla::Costing_Name(options.costing());
+  std::string routetype = valhalla::Costing_Enum_Name(options.costing());
   LOG_INFO("routetype: " + routetype);
 
   // Locations

--- a/src/worker.cc
+++ b/src/worker.cc
@@ -424,7 +424,7 @@ void parse_locations(const rapidjson::Document& doc,
         }
         auto preferred_side = rapidjson::get_optional<std::string>(r_loc, "/preferred_side");
         valhalla::Location::PreferredSide side;
-        if (preferred_side && PreferredSide_Parse(*preferred_side, &side)) {
+        if (preferred_side && PreferredSide_Enum_Parse(*preferred_side, &side)) {
           location->set_preferred_side(side);
         }
         auto search_cutoff = rapidjson::get_optional<unsigned int>(r_loc, "/search_cutoff");
@@ -500,7 +500,7 @@ void from_json(rapidjson::Document& doc, Options& options) {
 
   auto fmt = rapidjson::get_optional<std::string>(doc, "/format");
   Options::Format format;
-  if (fmt && Options_Format_Parse(*fmt, &format)) {
+  if (fmt && Options_Format_Enum_Parse(*fmt, &format)) {
     options.set_format(format);
   }
 
@@ -536,7 +536,7 @@ void from_json(rapidjson::Document& doc, Options& options) {
 
   auto dir_type = rapidjson::get_optional<std::string>(doc, "/directions_type");
   DirectionsType directions_type;
-  if (dir_type && DirectionsType_Parse(*dir_type, &directions_type)) {
+  if (dir_type && DirectionsType_Enum_Parse(*dir_type, &directions_type)) {
     options.set_directions_type(directions_type);
   }
 
@@ -638,7 +638,7 @@ void from_json(rapidjson::Document& doc, Options& options) {
   if (costing_str) {
     // try the string directly, some strings are keywords so add an underscore
     Costing costing;
-    if (valhalla::Costing_Parse(*costing_str, &costing)) {
+    if (valhalla::Costing_Enum_Parse(*costing_str, &costing)) {
       options.set_costing(costing);
     } else {
       throw valhalla_exception_t{125, "'" + *costing_str + "'"};
@@ -650,7 +650,7 @@ void from_json(rapidjson::Document& doc, Options& options) {
   for (const auto& costing : {auto_, auto_shorter, bicycle, bus, hov, motor_scooter, multimodal,
                               pedestrian, transit, truck, motorcycle, auto_data_fix, taxi}) {
     // Create the costing string
-    auto costing_str = valhalla::Costing_Name(costing);
+    auto costing_str = valhalla::Costing_Enum_Name(costing);
     // Create the costing options key
     const auto costing_options_key = "/costing_options/" + costing_str;
 
@@ -813,7 +813,7 @@ void from_json(rapidjson::Document& doc, Options& options) {
   auto shape_match_str = rapidjson::get_optional<std::string>(doc, "/shape_match");
   ShapeMatch shape_match;
   if (shape_match_str) {
-    if (valhalla::ShapeMatch_Parse(*shape_match_str, &shape_match)) {
+    if (valhalla::ShapeMatch_Enum_Parse(*shape_match_str, &shape_match)) {
       options.set_shape_match(shape_match);
     } else {
       throw valhalla_exception_t{445};
@@ -860,7 +860,7 @@ void from_json(rapidjson::Document& doc, Options& options) {
   // if specified, get the filter_action value in there
   auto filter_action_str = rapidjson::get_optional<std::string>(doc, "/filters/action");
   FilterAction filter_action;
-  if (filter_action_str && valhalla::FilterAction_Parse(*filter_action_str, &filter_action)) {
+  if (filter_action_str && valhalla::FilterAction_Enum_Parse(*filter_action_str, &filter_action)) {
     options.set_filter_action(filter_action);
   }
 
@@ -880,7 +880,7 @@ void from_json(rapidjson::Document& doc, Options& options) {
 
   // force these into the output so its obvious what we did to the user
   doc.AddMember({"language", allocator}, {options.language(), allocator}, allocator);
-  doc.AddMember({"format", allocator}, {valhalla::Options_Format_Name(options.format()), allocator},
+  doc.AddMember({"format", allocator}, {valhalla::Options_Format_Enum_Name(options.format()), allocator},
                 allocator);
 }
 
@@ -888,7 +888,7 @@ void from_json(rapidjson::Document& doc, Options& options) {
 
 namespace valhalla {
 
-bool Options_Action_Parse(const std::string& action, Options::Action* a) {
+bool Options_Action_Enum_Parse(const std::string& action, Options::Action* a) {
   static const std::unordered_map<std::string, Options::Action> actions{
       {"route", Options::route},
       {"locate", Options::locate},
@@ -908,7 +908,7 @@ bool Options_Action_Parse(const std::string& action, Options::Action* a) {
   return true;
 }
 
-const std::string& Options_Action_Name(const Options::Action action) {
+const std::string& Options_Action_Enum_Name(const Options::Action action) {
   static const std::string empty;
   static const std::unordered_map<int, std::string> actions{
       {Options::route, "route"},
@@ -926,7 +926,7 @@ const std::string& Options_Action_Name(const Options::Action action) {
   return i == actions.cend() ? empty : i->second;
 }
 
-bool Costing_Parse(const std::string& costing, Costing* c) {
+bool Costing_Enum_Parse(const std::string& costing, Costing* c) {
   static const std::unordered_map<std::string, Costing> costings{
       {"auto", Costing::auto_},
       {"auto_shorter", Costing::auto_shorter},
@@ -949,7 +949,7 @@ bool Costing_Parse(const std::string& costing, Costing* c) {
   return true;
 }
 
-const std::string& Costing_Name(const Costing costing) {
+const std::string& Costing_Enum_Name(const Costing costing) {
   static const std::string empty;
   static const std::unordered_map<int, std::string> costings{
       {Costing::auto_, "auto"},
@@ -970,7 +970,7 @@ const std::string& Costing_Name(const Costing costing) {
   return i == costings.cend() ? empty : i->second;
 }
 
-bool ShapeMatch_Parse(const std::string& match, ShapeMatch* s) {
+bool ShapeMatch_Enum_Parse(const std::string& match, ShapeMatch* s) {
   static const std::unordered_map<std::string, ShapeMatch> matches{
       {"edge_walk", ShapeMatch::edge_walk},
       {"map_snap", ShapeMatch::map_snap},
@@ -983,7 +983,7 @@ bool ShapeMatch_Parse(const std::string& match, ShapeMatch* s) {
   return true;
 }
 
-const std::string& ShapeMatch_Name(const ShapeMatch match) {
+const std::string& ShapeMatch_Enum_Name(const ShapeMatch match) {
   static const std::string empty;
   static const std::unordered_map<int, std::string> matches{
       {ShapeMatch::edge_walk, "edge_walk"},
@@ -994,7 +994,7 @@ const std::string& ShapeMatch_Name(const ShapeMatch match) {
   return i == matches.cend() ? empty : i->second;
 }
 
-bool Options_Format_Parse(const std::string& format, Options::Format* f) {
+bool Options_Format_Enum_Parse(const std::string& format, Options::Format* f) {
   static const std::unordered_map<std::string, Options::Format> formats{
       {"json", Options::json},
       {"gpx", Options::gpx},
@@ -1007,7 +1007,7 @@ bool Options_Format_Parse(const std::string& format, Options::Format* f) {
   return true;
 }
 
-const std::string& Options_Format_Name(const Options::Format match) {
+const std::string& Options_Format_Enum_Name(const Options::Format match) {
   static const std::string empty;
   static const std::unordered_map<int, std::string> formats{
       {Options::json, "json"},
@@ -1018,7 +1018,7 @@ const std::string& Options_Format_Name(const Options::Format match) {
   return i == formats.cend() ? empty : i->second;
 }
 
-const std::string& Options_Units_Name(const Options::Units unit) {
+const std::string& Options_Units_Enum_Name(const Options::Units unit) {
   static const std::string empty;
   static const std::unordered_map<int, std::string> units{
       {Options::kilometers, "kilometers"},
@@ -1028,7 +1028,7 @@ const std::string& Options_Units_Name(const Options::Units unit) {
   return i == units.cend() ? empty : i->second;
 }
 
-bool FilterAction_Parse(const std::string& action, FilterAction* a) {
+bool FilterAction_Enum_Parse(const std::string& action, FilterAction* a) {
   static const std::unordered_map<std::string, FilterAction> actions{
       {"exclude", FilterAction::exclude},
       {"include", FilterAction::include},
@@ -1040,7 +1040,7 @@ bool FilterAction_Parse(const std::string& action, FilterAction* a) {
   return true;
 }
 
-const std::string& FilterAction_Name(const FilterAction action) {
+const std::string& FilterAction_Enum_Name(const FilterAction action) {
   static const std::string empty;
   static const std::unordered_map<int, std::string> actions{
       {FilterAction::exclude, "exclude"},
@@ -1050,7 +1050,7 @@ const std::string& FilterAction_Name(const FilterAction action) {
   return i == actions.cend() ? empty : i->second;
 }
 
-bool DirectionsType_Parse(const std::string& dtype, DirectionsType* t) {
+bool DirectionsType_Enum_Parse(const std::string& dtype, DirectionsType* t) {
   static const std::unordered_map<std::string, DirectionsType> types{
       {"none", DirectionsType::none},
       {"maneuvers", DirectionsType::maneuvers},
@@ -1063,7 +1063,7 @@ bool DirectionsType_Parse(const std::string& dtype, DirectionsType* t) {
   return true;
 }
 
-bool PreferredSide_Parse(const std::string& pside, valhalla::Location::PreferredSide* p) {
+bool PreferredSide_Enum_Parse(const std::string& pside, valhalla::Location::PreferredSide* p) {
   static const std::unordered_map<std::string, valhalla::Location::PreferredSide> types{
       {"either", valhalla::Location::either},
       {"same", valhalla::Location::same},
@@ -1146,7 +1146,7 @@ void ParseApi(const http_request_t& request, valhalla::Api& api) {
 
   // set the action
   Options::Action action;
-  if (!request.path.empty() && Options_Action_Parse(request.path.substr(1), &action)) {
+  if (!request.path.empty() && Options_Action_Enum_Parse(request.path.substr(1), &action)) {
     options.set_action(action);
   }
 

--- a/src/worker.cc
+++ b/src/worker.cc
@@ -880,8 +880,8 @@ void from_json(rapidjson::Document& doc, Options& options) {
 
   // force these into the output so its obvious what we did to the user
   doc.AddMember({"language", allocator}, {options.language(), allocator}, allocator);
-  doc.AddMember({"format", allocator}, {valhalla::Options_Format_Enum_Name(options.format()), allocator},
-                allocator);
+  doc.AddMember({"format", allocator},
+                {valhalla::Options_Format_Enum_Name(options.format()), allocator}, allocator);
 }
 
 } // namespace

--- a/test/matrix.cc
+++ b/test/matrix.cc
@@ -177,7 +177,7 @@ void adjust_scores(Options& options) {
       }
 
       // subtract off the min score and cap at max so that path algorithm doesnt go too far
-      auto max_score = kMaxDistances.find(Costing_Name(options.costing()));
+      auto max_score = kMaxDistances.find(Costing_Enum_Name(options.costing()));
       for (auto* candidates : {location.mutable_path_edges(), location.mutable_filtered_edges()}) {
         for (auto& candidate : *candidates) {
           candidate.set_distance(candidate.distance() - minScore);

--- a/test/parse_request.cc
+++ b/test/parse_request.cc
@@ -227,8 +227,8 @@ void validate(const std::string& key,
   }
   if (pbf_value != expected_value) {
     throw std::runtime_error("Incorrect " + key +
-                             " ShapeMatch | expected_value=" + ShapeMatch_Name(expected_value) +
-                             " | found=" + ShapeMatch_Name(pbf_value));
+                             " ShapeMatch | expected_value=" + ShapeMatch_Enum_Name(expected_value) +
+                             " | found=" + ShapeMatch_Enum_Name(pbf_value));
   }
 }
 
@@ -240,9 +240,9 @@ void validate(const std::string& key,
     throw std::runtime_error("FilterAction value not found in pbf for key=" + key);
   }
   if (pbf_value != expected_value) {
-    throw std::runtime_error("Incorrect " + key +
-                             " FilterAction | expected_value=" + FilterAction_Name(expected_value) +
-                             " | found=" + FilterAction_Name(pbf_value));
+    throw std::runtime_error("Incorrect " + key + " FilterAction | expected_value=" +
+                             FilterAction_Enum_Name(expected_value) +
+                             " | found=" + FilterAction_Enum_Name(pbf_value));
   }
 }
 
@@ -340,11 +340,11 @@ std::string get_request_str(const std::string& key, const std::string& expected_
 }
 
 std::string get_request_str(const std::string& key, const ShapeMatch expected_value) {
-  return R"({")" + key + R"(":")" + ShapeMatch_Name(expected_value) + R"("})";
+  return R"({")" + key + R"(":")" + ShapeMatch_Enum_Name(expected_value) + R"("})";
 }
 
 std::string get_kv_str(const std::string& key, const valhalla::FilterAction value) {
-  return R"(")" + key + R"(":")" + FilterAction_Name(value) + R"(")";
+  return R"(")" + key + R"(":")" + FilterAction_Enum_Name(value) + R"(")";
 }
 
 std::string get_request_str(const std::string& parent_key,
@@ -394,7 +394,7 @@ Api get_request(const std::string& request_str, const Options::Action action) {
 // test parsing methods
 std::string get_costing_str(Costing costing) {
   // Create the costing string
-  auto costing_str = Costing_Name(costing);
+  auto costing_str = Costing_Enum_Name(costing);
   return costing_str;
 }
 

--- a/test/timedep_paths.cc
+++ b/test/timedep_paths.cc
@@ -77,7 +77,7 @@ void adjust_scores(Options& options) {
       }
 
       // subtract off the min score and cap at max so that path algorithm doesnt go too far
-      auto max_score = kMaxDistances.find(Costing_Name(options.costing()));
+      auto max_score = kMaxDistances.find(Costing_Enum_Name(options.costing()));
       for (auto* candidates : {location.mutable_path_edges(), location.mutable_filtered_edges()}) {
         for (auto& candidate : *candidates) {
           candidate.set_distance(candidate.distance() - minScore);

--- a/test/trivial_paths.cc
+++ b/test/trivial_paths.cc
@@ -76,7 +76,7 @@ void adjust_scores(Options& options) {
       }
 
       // subtract off the min score and cap at max so that path algorithm doesnt go too far
-      auto max_score = kMaxDistances.find(Costing_Name(options.costing()));
+      auto max_score = kMaxDistances.find(Costing_Enum_Name(options.costing()));
       for (auto* candidates : {location.mutable_path_edges(), location.mutable_filtered_edges()}) {
         for (auto& candidate : *candidates) {
           candidate.set_distance(candidate.distance() - minScore);

--- a/valhalla/sif/costfactory.h
+++ b/valhalla/sif/costfactory.h
@@ -50,7 +50,7 @@ public:
   cost_ptr_t Create(const Costing costing, const Options& options) const {
     auto itr = factory_funcs_.find(costing);
     if (itr == factory_funcs_.end()) {
-      auto costing_str = Costing_Name(costing);
+      auto costing_str = Costing_Enum_Name(costing);
       throw std::runtime_error("No costing method found for '" + costing_str + "'");
     }
     // create the cost using the function pointer

--- a/valhalla/worker.h
+++ b/valhalla/worker.h
@@ -16,19 +16,19 @@ namespace valhalla {
 
 // to use protobuflite we cant use descriptors which means we cant translate enums to strings
 // and so we reimplement the ones we use here
-bool Options_Action_Parse(const std::string& action, Options::Action* a);
-const std::string& Options_Action_Name(const Options::Action action);
-bool Costing_Parse(const std::string& costing, Costing* c);
-const std::string& Costing_Name(const Costing costing);
-bool ShapeMatch_Parse(const std::string& match, ShapeMatch* s);
-const std::string& ShapeMatch_Name(const ShapeMatch match);
-bool Options_Format_Parse(const std::string& format, Options::Format* f);
-const std::string& Options_Format_Name(const Options::Format match);
-const std::string& Options_Units_Name(const Options::Units unit);
-bool FilterAction_Parse(const std::string& action, FilterAction* a);
-const std::string& FilterAction_Name(const FilterAction action);
-bool DirectionsType_Parse(const std::string& dtype, DirectionsType* t);
-bool PreferredSide_Parse(const std::string& pside, valhalla::Location::PreferredSide* p);
+bool Options_Action_Enum_Parse(const std::string& action, Options::Action* a);
+const std::string& Options_Action_Enum_Name(const Options::Action action);
+bool Costing_Enum_Parse(const std::string& costing, Costing* c);
+const std::string& Costing_Enum_Name(const Costing costing);
+bool ShapeMatch_Enum_Parse(const std::string& match, ShapeMatch* s);
+const std::string& ShapeMatch_Enum_Name(const ShapeMatch match);
+bool Options_Format_Enum_Parse(const std::string& format, Options::Format* f);
+const std::string& Options_Format_Enum_Name(const Options::Format match);
+const std::string& Options_Units_Enum_Name(const Options::Units unit);
+bool FilterAction_Enum_Parse(const std::string& action, FilterAction* a);
+const std::string& FilterAction_Enum_Name(const FilterAction action);
+bool DirectionsType_Enum_Parse(const std::string& dtype, DirectionsType* t);
+bool PreferredSide_Enum_Parse(const std::string& pside, valhalla::Location::PreferredSide* p);
 
 const std::unordered_map<unsigned, std::string>
     error_codes{// loki project 1xx

